Promtail: Nil pointer reference, when parsing JSON logs

**Describe the bug**
I am trying to parse JSON logs with a pipeline in JSON. The JSON maps do not all have the same set of keys, e.g. the 'exception'-key is only present in a few logs. When I know want to extract the 'exception'- key-value pair and use it later on in the pipeline, e.g. to set it as a label, Promtail panics when trying to set the label, when the key did not exist in the JSON. Promatil gives me the following exception:

```
level=debug ts=2020-02-12T14:44:36.069455Z caller=pipeline.go:91 component=file_pipeline component=pipeline msg="processing pipeline" stage=0 name=json labels="{filename=\"/tmp/logs/log.json\", job=\"gerrit_error\"}" time=2020-02-12T15:44:36.06937+01:00 entry="{\"name\":\"test1\"}"
level=debug ts=2020-02-12T14:44:36.069974Z caller=pipeline.go:91 component=file_pipeline component=pipeline msg="processing pipeline" stage=1 name=labels labels="{filename=\"/tmp/logs/log.json\", job=\"gerrit_error\"}" time=2020-02-12T15:44:36.06937+01:00 entry="{\"name\":\"test1\"}"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xf8 pc=0x1a4beb0]

goroutine 259 [running]:
github.com/grafana/loki/pkg/logentry/stages.(*labelStage).Process(0xc00042fda0, 0xc0004eb710, 0xc0004eb740, 0xc0000ce6a0, 0xc000596370)
	/src/loki/pkg/logentry/stages/labels.go:71 +0x230
github.com/grafana/loki/pkg/logentry/stages.(*Pipeline).Process(0xc00009de80, 0xc0004eb710, 0xc0004eb740, 0xc0000ce6a0, 0xc000596370)
	/src/loki/pkg/logentry/stages/pipeline.go:93 +0x213
github.com/grafana/loki/pkg/logentry/stages.(*Pipeline).Wrap.func1(0xc0004eb710, 0xbf8921b504228090, 0x12c5bb147, 0x48be9e0, 0xc000584020, 0x10, 0x0, 0xffffffffffffffff)
	/src/loki/pkg/logentry/stages/pipeline.go:113 +0x101
github.com/grafana/loki/pkg/promtail/api.EntryHandlerFunc.Handle(0xc00042fde0, 0xc0004eb710, 0xbf8921b504228090, 0x12c5bb147, 0x48be9e0, 0xc000584020, 0x10, 0x203000, 0xc0004eb6e0)
	/src/loki/pkg/promtail/api/types.go:19 +0x65
github.com/grafana/loki/pkg/promtail/api.AddLabelsMiddleware.func1.1(0xc0004eb6e0, 0xbf8921b504228090, 0x12c5bb147, 0x48be9e0, 0xc000584020, 0x10, 0x0, 0x0)
	/src/loki/pkg/promtail/api/types.go:40 +0xa3
github.com/grafana/loki/pkg/promtail/api.EntryHandlerFunc.Handle(0xc0002460a0, 0xc0004eb6e0, 0xbf8921b504228090, 0x12c5bb147, 0x48be9e0, 0xc000584020, 0x10, 0xc0006e3d98, 0x0)
	/src/loki/pkg/promtail/api/types.go:19 +0x65
github.com/grafana/loki/pkg/promtail/api.AddLabelsMiddleware.func1.1(0xc0004eb6b0, 0xbf8921b504228090, 0x12c5bb147, 0x48be9e0, 0xc000584020, 0x10, 0x2fd8700, 0xc0004eb6b0)
	/src/loki/pkg/promtail/api/types.go:40 +0xa3
github.com/grafana/loki/pkg/promtail/api.EntryHandlerFunc.Handle(0xc000246300, 0xc0004eb6b0, 0xbf8921b504228090, 0x12c5bb147, 0x48be9e0, 0xc000584020, 0x10, 0x0, 0x0)
	/src/loki/pkg/promtail/api/types.go:19 +0x65
github.com/grafana/loki/pkg/promtail/targets.(*tailer).run(0xc0000c2720)
	/src/loki/pkg/promtail/targets/tailer.go:109 +0x47d
created by github.com/grafana/loki/pkg/promtail/targets.newTailer
	/src/loki/pkg/promtail/targets/tailer.go:74 +0x493
```

**To Reproduce**
Steps to reproduce the behavior:
1. Have a json log file:

```
{"name":"test1"}
{"name":"test2", "exception":"foo"}
```

2. Have a Promtail pipeline like:

```
  - json:
      expressions:
        name:
        exception:
  - labels:
      exception:
```

3. Start Promtail (1.3.0) to tail 'log.json'

**Expected behavior**
I would expect Promtail to skip non-existing key-value pairs in logs (maybe printing a warning). Thus, in the example above, I would get the label `exception` for the second, but not the first log entry, or at least have the label for the first entry set to `"nil"` or something similar.

**Environment:**
 - Infrastructure: laptop
 - Deployment tool: binary

**Screenshots, Promtail config, or terminal output**
see description above.

--- a/pkg/logentry/stages/labels.go
+++ b/pkg/logentry/stages/labels.go
@@ -68,7 +68,7 @@ func (l *labelStage) Process(labels model.LabelSet, extracted map[string]interfa
 			s, err := getString(lValue)
 			if err != nil {
 				if Debug {
-					level.Debug(l.logger).Log("msg", "failed to convert extracted label value to string", "err", err, "type", reflect.TypeOf(lValue).String())
+					level.Debug(l.logger).Log("msg", "failed to convert extracted label value to string", "err", err, "type", reflect.TypeOf(lValue))
 				}
 				continue
 			}

--- a/pkg/logentry/stages/labels_test.go
+++ b/pkg/logentry/stages/labels_test.go
@@ -1,12 +1,15 @@
 package stages
 
 import (
+	"bytes"
 	"errors"
 	"fmt"
+	"strings"
 	"testing"
 	"time"
 
 	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/common/model"
 	"github.com/stretchr/testify/assert"
@@ -31,6 +34,13 @@ var testLabelsLogLine = `
 	"level" : "WARN"
 }
 `
+var testLabelsLogLineWithMissingKey = `
+{
+	"time":"2012-11-01T22:08:41+00:00",
+	"app":"loki",
+	"component": ["parser","type"]
+}
+`
 
 func TestLabelsPipeline_Labels(t *testing.T) {
 	pl, err := NewPipeline(util.Logger, loadConfig(testLabelsYaml), nil, prometheus.DefaultRegisterer)
@@ -47,6 +57,26 @@ func TestLabelsPipeline_Labels(t *testing.T) {
 	extracted := map[string]interface{}{}
 	pl.Process(lbls, extracted, &ts, &entry)
 	assert.Equal(t, expectedLbls, lbls)
+}
+
+func TestLabelsPipelineWithMissingKey_Labels(t *testing.T) {
+	var buf bytes.Buffer
+	w := log.NewSyncWriter(&buf)
+	logger := log.NewLogfmtLogger(w)
+	pl, err := NewPipeline(logger, loadConfig(testLabelsYaml), nil, prometheus.DefaultRegisterer)
+	if err != nil {
+		t.Fatal(err)
+	}
+	lbls := model.LabelSet{}
+	Debug = true
+	ts := time.Now()
+	entry := testLabelsLogLineWithMissingKey
+	extracted := map[string]interface{}{}
+	pl.Process(lbls, extracted, &ts, &entry)
+	expectedLog := "level=debug msg=\"failed to convert extracted label value to string\" err=\"Can't convert <nil> to string\" type=null"
+	if !(strings.Contains(buf.String(), expectedLog)) {
+		t.Errorf("\nexpected: %s\n+actual: %s", expectedLog, buf.String())
+	}
 }
 
 var (

--- a/pkg/logentry/stages/metrics.go
+++ b/pkg/logentry/stages/metrics.go
@@ -179,7 +179,7 @@ func (m *metricStage) recordCounter(name string, counter *metric.Counters, label
 			if Debug {
 				level.Debug(m.logger).Log("msg", "failed to convert extracted value to string, "+
 					"can't perform value comparison", "metric", name, "err",
-					fmt.Sprintf("can't convert %v to string", reflect.TypeOf(v).String()))
+					fmt.Sprintf("can't convert %v to string", reflect.TypeOf(v)))
 			}
 			return
 		}
@@ -212,7 +212,7 @@ func (m *metricStage) recordGauge(name string, gauge *metric.Gauges, labels mode
 			if Debug {
 				level.Debug(m.logger).Log("msg", "failed to convert extracted value to string, "+
 					"can't perform value comparison", "metric", name, "err",
-					fmt.Sprintf("can't convert %v to string", reflect.TypeOf(v).String()))
+					fmt.Sprintf("can't convert %v to string", reflect.TypeOf(v)))
 			}
 			return
 		}
@@ -265,7 +265,7 @@ func (m *metricStage) recordHistogram(name string, histogram *metric.Histograms,
 			if Debug {
 				level.Debug(m.logger).Log("msg", "failed to convert extracted value to string, "+
 					"can't perform value comparison", "metric", name, "err",
-					fmt.Sprintf("can't convert %v to string", reflect.TypeOf(v).String()))
+					fmt.Sprintf("can't convert %v to string", reflect.TypeOf(v)))
 			}
 			return
 		}

--- a/pkg/logentry/stages/metrics_test.go
+++ b/pkg/logentry/stages/metrics_test.go
@@ -1,11 +1,13 @@
 package stages
 
 import (
+	"bytes"
 	"strings"
 	"testing"
 	"time"
 
 	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
 	"github.com/pkg/errors"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/testutil"
@@ -69,6 +71,14 @@ var testMetricLogLine2 = `
 	"level" : "WARN"
 }
 `
+var testMetricLogLineWithMissingKey = `
+{
+	"time":"2012-11-01T22:08:41+00:00",
+	"payload": 20,
+	"component": ["parser","type"],
+	"level" : "WARN"
+}
+`
 
 const expectedMetrics = `# HELP my_promtail_custom_loki_count uhhhhhhh
 # TYPE my_promtail_custom_loki_count counter
@@ -106,6 +116,26 @@ func TestMetricsPipeline(t *testing.T) {
 	if err := testutil.GatherAndCompare(registry,
 		strings.NewReader(expectedMetrics)); err != nil {
 		t.Fatalf("missmatch metrics: %v", err)
+	}
+}
+
+func TestPipelineWithMissingKey_Metrics(t *testing.T) {
+	var buf bytes.Buffer
+	w := log.NewSyncWriter(&buf)
+	logger := log.NewLogfmtLogger(w)
+	pl, err := NewPipeline(logger, loadConfig(testMetricYaml), nil, prometheus.DefaultRegisterer)
+	if err != nil {
+		t.Fatal(err)
+	}
+	lbls := model.LabelSet{}
+	Debug = true
+	ts := time.Now()
+	entry := testMetricLogLineWithMissingKey
+	extracted := map[string]interface{}{}
+	pl.Process(lbls, extracted, &ts, &entry)
+	expectedLog := "level=debug msg=\"failed to convert extracted value to string, can't perform value comparison\" metric=bloki_count err=\"can't convert <nil> to string\""
+	if !(strings.Contains(buf.String(), expectedLog)) {
+		t.Errorf("\nexpected: %s\n+actual: %s", expectedLog, buf.String())
 	}
 }
 

--- a/pkg/logentry/stages/output.go
+++ b/pkg/logentry/stages/output.go
@@ -65,7 +65,7 @@ func (o *outputStage) Process(labels model.LabelSet, extracted map[string]interf
 		s, err := getString(v)
 		if err != nil {
 			if Debug {
-				level.Debug(o.logger).Log("msg", "extracted output could not be converted to a string", "err", err, "type", reflect.TypeOf(v).String())
+				level.Debug(o.logger).Log("msg", "extracted output could not be converted to a string", "err", err, "type", reflect.TypeOf(v))
 			}
 			return
 		}

--- a/pkg/logentry/stages/output_test.go
+++ b/pkg/logentry/stages/output_test.go
@@ -1,10 +1,13 @@
 package stages
 
 import (
+	"bytes"
+	"strings"
 	"testing"
 	"time"
 
 	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
 	"github.com/pkg/errors"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/common/model"
@@ -30,6 +33,15 @@ var testOutputLogLine = `
 	"message" : "this is a log line"
 }
 `
+var testOutputLogLineWithMissingKey = `
+{
+	"time":"2012-11-01T22:08:41+00:00",
+	"app":"loki",
+	"component": ["parser","type"],
+	"level" : "WARN",
+	"nested" : {"child":"value"}
+}
+`
 
 func TestPipeline_Output(t *testing.T) {
 	pl, err := NewPipeline(util.Logger, loadConfig(testOutputYaml), nil, prometheus.DefaultRegisterer)
@@ -42,6 +54,26 @@ func TestPipeline_Output(t *testing.T) {
 	extracted := map[string]interface{}{}
 	pl.Process(lbls, extracted, &ts, &entry)
 	assert.Equal(t, "this is a log line", entry)
+}
+
+func TestPipelineWithMissingKey_Output(t *testing.T) {
+	var buf bytes.Buffer
+	w := log.NewSyncWriter(&buf)
+	logger := log.NewLogfmtLogger(w)
+	pl, err := NewPipeline(logger, loadConfig(testOutputYaml), nil, prometheus.DefaultRegisterer)
+	if err != nil {
+		t.Fatal(err)
+	}
+	lbls := model.LabelSet{}
+	Debug = true
+	ts := time.Now()
+	entry := testOutputLogLineWithMissingKey
+	extracted := map[string]interface{}{}
+	pl.Process(lbls, extracted, &ts, &entry)
+	expectedLog := "level=debug msg=\"extracted output could not be converted to a string\" err=\"Can't convert <nil> to string\" type=null"
+	if !(strings.Contains(buf.String(), expectedLog)) {
+		t.Errorf("\nexpected: %s\n+actual: %s", expectedLog, buf.String())
+	}
 }
 
 func TestOutputValidation(t *testing.T) {

--- a/pkg/logentry/stages/regex.go
+++ b/pkg/logentry/stages/regex.go
@@ -99,7 +99,7 @@ func (r *regexStage) Process(labels model.LabelSet, extracted map[string]interfa
 		value, err := getString(extracted[*r.cfg.Source])
 		if err != nil {
 			if Debug {
-				level.Debug(r.logger).Log("msg", "failed to convert source value to string", "source", *r.cfg.Source, "err", err, "type", reflect.TypeOf(extracted[*r.cfg.Source]).String())
+				level.Debug(r.logger).Log("msg", "failed to convert source value to string", "source", *r.cfg.Source, "err", err, "type", reflect.TypeOf(extracted[*r.cfg.Source]))
 			}
 			return
 		}

--- a/pkg/logentry/stages/regex_test.go
+++ b/pkg/logentry/stages/regex_test.go
@@ -1,11 +1,14 @@
 package stages
 
 import (
+	"bytes"
 	"reflect"
+	"strings"
 	"testing"
 	"time"
 
 	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
 	"github.com/pkg/errors"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/common/model"
@@ -26,6 +29,23 @@ pipeline_stages:
 - regex:
     expression: "^HTTP\\/(?P<protocol_version>[0-9\\.]+)$"
     source:     "protocol"
+`
+var testRegexYamlSourceWithMissingKey = `
+pipeline_stages:
+- json:
+    expressions:
+      time:
+- regex:
+    expression: "^(?P<year>\\d+)"
+    source:     "time"
+`
+
+var testRegexLogLineWithMissingKey = `
+{
+	"app":"loki",
+	"component": ["parser","type"],
+	"level": "WARN"
+}
 `
 
 var testRegexLogLine = `11.11.11.11 - frank [25/Jan/2000:14:00:01 -0500] "GET /1986.js HTTP/1.1" 200 932 "-" "Mozilla/5.0 (Windows; U; Windows NT 5.1; de; rv:1.9.1.7) Gecko/20091221 Firefox/3.5.7 GTB6"`
@@ -93,6 +113,26 @@ func TestPipeline_Regex(t *testing.T) {
 			pl.Process(lbls, extracted, &ts, &entry)
 			assert.Equal(t, testData.expectedExtract, extracted)
 		})
+	}
+}
+
+func TestPipelineWithMissingKey_Regex(t *testing.T) {
+	var buf bytes.Buffer
+	w := log.NewSyncWriter(&buf)
+	logger := log.NewLogfmtLogger(w)
+	pl, err := NewPipeline(logger, loadConfig(testRegexYamlSourceWithMissingKey), nil, prometheus.DefaultRegisterer)
+	if err != nil {
+		t.Fatal(err)
+	}
+	lbls := model.LabelSet{}
+	Debug = true
+	ts := time.Now()
+	entry := testRegexLogLineWithMissingKey
+	extracted := map[string]interface{}{}
+	pl.Process(lbls, extracted, &ts, &entry)
+	expectedLog := "level=debug component=stage type=regex msg=\"failed to convert source value to string\" source=time err=\"Can't convert <nil> to string\" type=null"
+	if !(strings.Contains(buf.String(), expectedLog)) {
+		t.Errorf("\nexpected: %s\n+actual: %s", expectedLog, buf.String())
 	}
 }
 

--- a/pkg/logentry/stages/template.go
+++ b/pkg/logentry/stages/template.go
@@ -88,7 +88,7 @@ func (o *templateStage) Process(labels model.LabelSet, extracted map[string]inte
 		s, err := getString(v)
 		if err != nil {
 			if Debug {
-				level.Debug(o.logger).Log("msg", "extracted template could not be converted to a string", "err", err, "type", reflect.TypeOf(v).String())
+				level.Debug(o.logger).Log("msg", "extracted template could not be converted to a string", "err", err, "type", reflect.TypeOf(v))
 			}
 			return
 		}

--- a/pkg/logentry/stages/template_test.go
+++ b/pkg/logentry/stages/template_test.go
@@ -1,11 +1,14 @@
 package stages
 
 import (
+	"bytes"
 	"errors"
+	"strings"
 	"testing"
 	"time"
 
 	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/common/model"
 	"github.com/stretchr/testify/assert"
@@ -42,6 +45,15 @@ var testTemplateLogLine = `
 	"message" : "this is a log line"
 }
 `
+var testTemplateLogLineWithMissingKey = `
+{
+	"time":"2012-11-01T22:08:41+00:00",
+	"component": ["parser","type"],
+	"level" : "WARN",
+	"nested" : {"child":"value"},
+	"message" : "this is a log line"
+}
+`
 
 func TestPipeline_Template(t *testing.T) {
 	pl, err := NewPipeline(util.Logger, loadConfig(testTemplateYaml), nil, prometheus.DefaultRegisterer)
@@ -59,6 +71,26 @@ func TestPipeline_Template(t *testing.T) {
 	extracted := map[string]interface{}{}
 	pl.Process(lbls, extracted, &ts, &entry)
 	assert.Equal(t, expectedLbls, lbls)
+}
+
+func TestPipelineWithMissingKey_Temaplate(t *testing.T) {
+	var buf bytes.Buffer
+	w := log.NewSyncWriter(&buf)
+	logger := log.NewLogfmtLogger(w)
+	pl, err := NewPipeline(logger, loadConfig(testTemplateYaml), nil, prometheus.DefaultRegisterer)
+	if err != nil {
+		t.Fatal(err)
+	}
+	lbls := model.LabelSet{}
+	Debug = true
+	ts := time.Now()
+	entry := testTemplateLogLineWithMissingKey
+	extracted := map[string]interface{}{}
+	pl.Process(lbls, extracted, &ts, &entry)
+	expectedLog := "level=debug msg=\"extracted template could not be converted to a string\" err=\"Can't convert <nil> to string\" type=null"
+	if !(strings.Contains(buf.String(), expectedLog)) {
+		t.Errorf("\nexpected: %s\n+actual: %s", expectedLog, buf.String())
+	}
 }
 
 func TestTemplateValidation(t *testing.T) {

--- a/pkg/logentry/stages/tenant.go
+++ b/pkg/logentry/stages/tenant.go
@@ -97,7 +97,7 @@ func (s *tenantStage) getTenantFromSourceField(extracted map[string]interface{})
 	tenantID, err := getString(value)
 	if err != nil {
 		if Debug {
-			level.Debug(s.logger).Log("msg", "failed to convert value to string", "err", err, "type", reflect.TypeOf(value).String())
+			level.Debug(s.logger).Log("msg", "failed to convert value to string", "err", err, "type", reflect.TypeOf(value))
 		}
 		return ""
 	}

--- a/pkg/logentry/stages/tenant_test.go
+++ b/pkg/logentry/stages/tenant_test.go
@@ -1,16 +1,58 @@
 package stages
 
 import (
+	"bytes"
+	"strings"
 	"testing"
 	"time"
 
 	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
 	"github.com/grafana/loki/pkg/promtail/client"
 	lokiutil "github.com/grafana/loki/pkg/util"
+	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/common/model"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
+
+var testTenantYaml = `
+pipeline_stages:
+- json:
+    expressions:
+      customer_id:
+- tenant:
+    source: customer_id
+`
+
+var testTenantLogLineWithMissingKey = `
+{
+	"time":"2012-11-01T22:08:41+00:00",
+	"app":"loki",
+	"component": ["parser","type"],
+	"level" : "WARN"
+}
+`
+
+func TestPipelineWithMissingKey_Tenant(t *testing.T) {
+	var buf bytes.Buffer
+	w := log.NewSyncWriter(&buf)
+	logger := log.NewLogfmtLogger(w)
+	pl, err := NewPipeline(logger, loadConfig(testTenantYaml), nil, prometheus.DefaultRegisterer)
+	if err != nil {
+		t.Fatal(err)
+	}
+	lbls := model.LabelSet{}
+	Debug = true
+	ts := time.Now()
+	entry := testTenantLogLineWithMissingKey
+	extracted := map[string]interface{}{}
+	pl.Process(lbls, extracted, &ts, &entry)
+	expectedLog := "level=debug msg=\"failed to convert value to string\" err=\"Can't convert <nil> to string\" type=null"
+	if !(strings.Contains(buf.String(), expectedLog)) {
+		t.Errorf("\nexpected: %s\n+actual: %s", expectedLog, buf.String())
+	}
+}
 
 func TestTenantStage_Validation(t *testing.T) {
 	t.Parallel()

--- a/pkg/logentry/stages/timestamp.go
+++ b/pkg/logentry/stages/timestamp.go
@@ -165,7 +165,7 @@ func (ts *timestampStage) parseTimestampFromSource(extracted map[string]interfac
 	s, err := getString(v)
 	if err != nil {
 		if Debug {
-			level.Debug(ts.logger).Log("msg", ErrTimestampConversionFailed, "err", err, "type", reflect.TypeOf(v).String())
+			level.Debug(ts.logger).Log("msg", ErrTimestampConversionFailed, "err", err, "type", reflect.TypeOf(v))
 		}
 
 		return nil, errors.New(ErrTimestampConversionFailed)

--- a/pkg/logentry/stages/timestamp_test.go
+++ b/pkg/logentry/stages/timestamp_test.go
@@ -1,12 +1,14 @@
 package stages
 
 import (
+	"bytes"
 	"fmt"
 	"strings"
 	"testing"
 	"time"
 
 	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
 	lokiutil "github.com/grafana/loki/pkg/util"
 	"github.com/pkg/errors"
 	"github.com/prometheus/client_golang/prometheus"
@@ -34,6 +36,14 @@ var testTimestampLogLine = `
 }
 `
 
+var testTimestampLogLineWithMissingKey = `
+{
+	"app":"loki",
+	"component": ["parser","type"],
+	"level" : "WARN"
+}
+`
+
 func TestTimestampPipeline(t *testing.T) {
 	pl, err := NewPipeline(util.Logger, loadConfig(testTimestampYaml), nil, prometheus.DefaultRegisterer)
 	if err != nil {
@@ -52,6 +62,26 @@ var (
 	validLocationString   = "America/New_York"
 	validLocation, _      = time.LoadLocation(validLocationString)
 )
+
+func TestPipelineWithMissingKey_Timestamp(t *testing.T) {
+	var buf bytes.Buffer
+	w := log.NewSyncWriter(&buf)
+	logger := log.NewLogfmtLogger(w)
+	pl, err := NewPipeline(logger, loadConfig(testTimestampYaml), nil, prometheus.DefaultRegisterer)
+	if err != nil {
+		t.Fatal(err)
+	}
+	lbls := model.LabelSet{}
+	Debug = true
+	ts := time.Now()
+	entry := testTimestampLogLineWithMissingKey
+	extracted := map[string]interface{}{}
+	pl.Process(lbls, extracted, &ts, &entry)
+	expectedLog := fmt.Sprintf("level=debug msg=\"%s\" err=\"Can't convert <nil> to string\" type=null", ErrTimestampConversionFailed)
+	if !(strings.Contains(buf.String(), expectedLog)) {
+		t.Errorf("\nexpected: %s\n+actual: %s", expectedLog, buf.String())
+	}
+}
 
 func TestTimestampValidation(t *testing.T) {
 	tests := map[string]struct {

